pkg/kubelet/cm/memorymanager/policy_static.go

```diff
@@ -788,3 +830,102 @@ func (p *staticPolicy) GetAllocatableMemory(s state.State) []state.Block {
 	}
 	return allocatableMemory
 }
+
+func (p *staticPolicy) updatePodReusableMemory(pod *v1.Pod, container *v1.Container, memoryBlocks []state.Block) {
+	podUID := string(pod.UID)
+
+	// If pod entries to m.initContainersReusableMemory other than the current pod exist, delete them.
+	for uid := range p.initContainersReusableMemory {
+		if podUID != uid {
+			delete(p.initContainersReusableMemory, podUID)
+		}
+	}
+
+	if isInitContainer(pod, container) {
+		if _, ok := p.initContainersReusableMemory[podUID]; !ok {
+			p.initContainersReusableMemory[podUID] = map[string]map[v1.ResourceName]uint64{}
+		}
+
+		for _, block := range memoryBlocks {
+			blockBitMask, _ := bitmask.NewBitMask(block.NUMAAffinity...)
+			blockBitMaskString := blockBitMask.String()
+
+			if _, ok := p.initContainersReusableMemory[podUID][blockBitMaskString]; !ok {
+				p.initContainersReusableMemory[podUID][blockBitMaskString] = map[v1.ResourceName]uint64{}
+			}
+
+			if blockReusableMemory := p.initContainersReusableMemory[podUID][blockBitMaskString][block.Type]; block.Size > blockReusableMemory {
+				p.initContainersReusableMemory[podUID][blockBitMaskString][block.Type] = block.Size
+			}
+		}
+
+		return
+	}
+
+	// update re-usable memory once it used by the app container
+	for _, block := range memoryBlocks {
+		blockBitMask, _ := bitmask.NewBitMask(block.NUMAAffinity...)
+		if podReusableMemory := p.getPodReusableMemory(pod, blockBitMask, block.Type); podReusableMemory != 0 {
+			if block.Size >= podReusableMemory {
+				p.initContainersReusableMemory[podUID][blockBitMask.String()][block.Type] = 0
+			} else {
+				p.initContainersReusableMemory[podUID][blockBitMask.String()][block.Type] -= block.Size
+			}
+		}
+	}
+}
+
+func (p *staticPolicy) updateInitContainersMemoryBlocks(s state.State, pod *v1.Pod) {
+	podUID := string(pod.UID)
+	updatedPodReusableMemory := map[string]map[v1.ResourceName]uint64{}
+	for numaAffinity, resourceReusable := range p.initContainersReusableMemory[podUID] {
+		if _, ok := updatedPodReusableMemory[numaAffinity]; !ok {
+			updatedPodReusableMemory[numaAffinity] = map[v1.ResourceName]uint64{}
+		}
+
+		for resourceName, size := range resourceReusable {
+			updatedPodReusableMemory[numaAffinity][resourceName] = size
+		}
+	}
+
+	for i := len(pod.Spec.InitContainers) - 1; i >= 0; i-- {
```

Comment:
Is there any particular reason to walk backwards?

Reply:
The init containers started in the order specified under `pod.Spec.InitContainers`, so first, we want to re-use the memory allocated for the first init container, so the update of the re-usable memory should go backward.

Reply:
Changed the logic, no need anymore.